### PR TITLE
G-API: Support CUDA & TensoRT Execution Providers for ONNXRT Backend

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by TolyaTalamanov** [comment]
Based on this PR:
https://github.com/opencv/opencv/pull/24045

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp [hunk 1]
```diff
@@ -32,6 +32,56 @@ namespace onnx {
  */
 namespace ep {
 
+/**
+ * @brief This structure provides functions
+ * that fill inference options for CUDA OpenVINO Execution Provider.
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/infer/onnx.hpp:37]: Just `CUDA`

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp [hunk 2]
```diff
@@ -32,6 +32,56 @@ namespace onnx {
...
+    explicit CUDA(const int dev_id)
+        : device_id(dev_id) {
+    }
+
+    int device_id;
+};
+
+/**
+ * @brief This structure provides functions
+ * that fill inference options for TensorRT OpenVINO Execution Provider.
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/infer/onnx.hpp:62]: Just `TensorRT` (copy-paste from `OpenVINO`)

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp [hunk 3]
```diff
@@ -431,6 +485,34 @@ template<typename Net> class Params {
...
+        desc.execution_providers.emplace_back(std::move(ep));
+        return *this;
+    }
+
+    /** @brief Adds execution provider for runtime.
+
+    The function is used to add ONNX Runtime TensorRT Execution Provider options.
+
+    @param ep TensorRT Execution Provider options.
+    @see cv::gapi::onnx::ep::CUDA.
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/infer/onnx.hpp:507]: `TensortRT`

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp [hunk 4]
```diff
@@ -145,9 +145,39 @@ class ONNXCompiled {
...
+         ss << "ONNX Backend: Failed to enable CUDA"
+            << " Execution Provider: " << e.what();
+         cv::util::throw_error(std::runtime_error(ss.str()));
+     }
+}
+
+static void addTensorRTExecutionProvider(Ort::SessionOptions *session_options,
+                                         const cv::gapi::onnx::ep::TensorRT &trt_ep) {
+     OrtTensorRTProviderOptions options{};
+     options.device_id = trt_ep.device_id;
```
Comment (TolyaTalamanov) [inline modules/gapi/src/backends/onnx/gonnxbackend.cpp:166]: Consider more options:
```
options.trt_max_workspace_size = 2147483648;
options.trt_max_partition_iterations = 10;
options.trt_min_subgraph_size = 5;
options.trt_fp16_enable = 1;
options.trt_engine_cache_enable = 1;
options.trt_engine_cache_path = "ort_tensorrt_cache";
options.trt_dump_subgraphs = 1;  
```

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp (hunk 1 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/infer/onnx.hpp:37]: Done

File: modules/gapi/include/opencv2/gapi/infer/onnx.hpp (hunk 2 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/infer/onnx.hpp:62]: Done

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp (hunk 4 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/onnx/gonnxbackend.cpp:166]: Well technically they can be provided via env variable:
https://onnxruntime.ai/docs/execution-providers/TensorRT-ExecutionProvider.html#execution-provider-options

Not clear at this point which of them is important to expose to API

**Review by dmatveev — APPROVED** [review]

**Comment by TolyaTalamanov** [comment]
@asmorkalov We're ready to merge this

**Pull request merged**